src/profitability.py

```diff
@@ -72,11 +83,236 @@ def profitability(self, fiat: fiat, hashrate: int, power_consumption: int, elect
         data["profitability"] = 100 * (data["expected_fiat_income_s"] - data["mining_cost_s"]) / data["mining_cost_s"] if data["mining_cost_s"] != 0 else math.inf
         data["breakeven_efficiency"] = (data["difficulty"] * electricity_cost) / (data["price"] * data["reward"] * 1000 * 3600 * (1 - pool_fee))
         return data
+    
+    def get_info(self) -> dict:
+        now = datetime.now()
+        json_req = { "jsonrpc": "2.0", "id": "0", "method": "get_info"}
+        try:
+            data = requests.get(f"{self.node_url}/json_rpc", json=json_req).json()
+        except Exception as e:
+            print("Error!", e)
+            pass
+        else:
+            self._height = data["result"]["height"]
+            self._height_last_fetched = now
+            self._difficulty = data["result"]["difficulty"]
+            self._difficulty_last_fetched = now
+        return data["result"]
+    
+    def _request_headers_range(self, start_height:int, end_height:int) -> pd.DataFrame:
+        json_req = {
+            "jsonrpc": "2.0",
+            "id": "0",
+            "method": "get_block_headers_range",
+            "params": {"start_height": start_height, "end_height": end_height}
+        }
+        try:
+            data = requests.post(f"{self.node_url}/json_rpc", json=json_req).json()
+            if "result" in data:
+                result = pd.DataFrame(data["result"]["headers"], columns=["height", "timestamp", "difficulty", "reward"])
+                result.set_index("height", inplace=True)
+                return result
+            elif "error" in data:
+                raise Exception(f"Error while attempting to fetch headers {start_height}-{end_height} from {self.node_url}\n{data['error']}")
+            else:
+                raise Exception(f"Something went terribly wrong while attempting to fetch headers {start_height}-{end_height} from {self.node_url}\n{data}")
+        except Exception as e:
+            print(f"Error while attempting to fetch headers {start_height}-{end_height} from {self.node_url}")
+            print(e)
+    
+    def _request_headers_batcher(self, start_height:int, end_height:int, batch_size:int=1000) -> Tuple[pd.DataFrame, bool]:
+        # if (end_height - start_height) >= batch_size, send a batch request then
+        # evaluate again with start_height = start_height + batch_size
+        # if it's false request the last range from start_height to end_height
+        results = []
+        try:
+            print(f"Downloading headers from {start_height} to {end_height}")
+            while (end_height - start_height >= batch_size):
```

Comment:
This loop has to have any kind of `max_iterations` fuse, where the `max_iterations` can be a very large number.